File: impls/bun/step0_repl.ts

const READ = (input: string) => input;

const EVAL = (input: string) => input;
const PRINT = (input: string) => input;

const prompt = "user> ";

const rep = (input: string) => PRINT(EVAL(READ(input)))

const repl = async () => {
    console.write(prompt);
    for await (const line of console) {
        console.log(rep(line));
        console.write(prompt);
    }
}
repl()

File: impls/elixir2/lib/mix/tasks/step0_repl.ex

defmodule Mix.Tasks.Step0Repl do
  def run(_), do: loop()

  defp loop do
    IO.write(:stdio, "user> ")
    IO.read(:stdio, :line)
      |> read_eval_print
      |> IO.puts

    loop()
  end

  defp read(input) do
    input
  end

  defp eval(input) do
    input
  end

  defp print(input) do
    input
  end

  defp read_eval_print(:eof), do: exit(:normal)
  defp read_eval_print(line) do
    read(line)
      |> eval
      |> print
  end

end

